### PR TITLE
Capturing information at moment of crash

### PR DESCRIPTION
**Pull request by ShreyaA**
Changes to create description file at moment of crash

Please have a look at our [guidelines for contributions](https://github.com/microsoft/appcenter-sdk-android/blob/develop/CONTRIBUTING.md) and consider the following before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [ ] Are tests passing locally?
* [ ] Are the files formatted correctly?
* [ ] Did you add unit tests?
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.

## Description

A few sentences describing the overall goals of the pull request.

## Related PRs or issues

List related PRs and other issues.

## Misc

Add what's missing, notes on what you tested, additional thoughts or questions.

### CODE REVIEW

**Comment by russelarms** [comment]
Hi @ShreyaA, thanks for the contribution!
As I understand the main goal of this PR is to add an ability to provide a custom description in the moment of a crash, e.g. logcat logs?
This behavior can be achieved by using a custom `UncaughtExceptionHandler` as it was described [here](https://github.com/microsoft/appcenter-sdk-android/issues/1397#issuecomment-608505531).
So at first, you would need to save any information you want inside your custom exception handler and
then you can pass this info as attachment by overriding `CrashesListener.getErrorAttachments()` method.

**Comment by ShreyaA** [comment]
> Hi @ShreyaA, thanks for the contribution!
> As I understand the main goal of this PR is to add an ability to provide a custom description in the moment of a crash, e.g. logcat logs?
> This behavior can be achieved by using a custom `UncaughtExceptionHandler` as it was described [here](https://github.com/microsoft/appcenter-sdk-android/issues/1397#issuecomment-608505531).
> So at first, you would need to save any information you want inside your custom exception handler and
> then you can pass this info as attachment by overriding `CrashesListener.getErrorAttachments()` method.

Hi @russelarms 
Thanks for sharing this! Does filename mean this file will get stored with the report id we see on AppCenter?
Will this work for both Java and Native crashes?

**Comment by ShreyaA** [comment]
Following the method described here: https://github.com/microsoft/appcenter-sdk-android/issues/1397
We will leverage the UUID being returned in this manner: 

```

 @Override
    public void uncaughtException(Thread thread, Throwable exception) {
       //mDefaultUncaughtExceptionHandler is com.microsoft.appcenter.crashes.UncaughtExceptionHandler here
        UUID reportUUID = Crashes.getInstance().saveUncaughtException(thread, exception);
       //save relevant files with UUID as name + custom extension
       mDefaultUncaughtExceptionHandler.uncaughtException(thread, exception);
}

```

File: sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/CrashesListener.java [hunk 1]
```diff
@@ -62,4 +62,5 @@
      * @param report The crash report that was sent successfully.
      */
     void onSendingSucceeded(ErrorReport report);
+
```
Comment (MatkovIvan) [inline sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/CrashesListener.java:65]:
```suggestion
```

File: sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java [hunk 2]
```diff
@@ -839,7 +840,6 @@ private void processPendingErrors() {
                     }
                 } catch (JSONException e) {
                     AppCenterLog.error(LOG_TAG, "Error parsing error log. Deleting invalid file: " + logFile, e);
-
                     //noinspection ResultOfMethodCallIgnored
```
Comment (MatkovIvan) [inline sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java:843]:
```suggestion

                    //noinspection ResultOfMethodCallIgnored
```

File: sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java [hunk 3]
```diff
@@ -341,6 +341,7 @@ public static void notifyUserConfirmation(@UserConfirmationDef int userConfirmat
         return getInstance().getInstanceLastSessionCrashReport();
     }
 
+
```
Comment (MatkovIvan) [inline sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java:344]:
```suggestion
```

File: sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/AbstractCrashesListener.java [hunk 4]
```diff
@@ -7,7 +7,6 @@
 
 import com.microsoft.appcenter.crashes.ingestion.models.ErrorAttachmentLog;
 import com.microsoft.appcenter.crashes.model.ErrorReport;
-
 /**
```
Comment (MatkovIvan) [inline sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/AbstractCrashesListener.java:10]:
```suggestion

/**
```

**Review by MatkovIvan — COMMENTED** [review]
A few reverting whitespace suggestions

File: sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java [hunk 5]
```diff
@@ -1097,14 +1097,16 @@ void setLogSerializer(LogSerializer logSerializer) {
      * @param thread    thread where crash occurred.
```
Comment (dhei) [inline sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java:1097]: Thanks for the PR.

To follow the convention, please add this before the `@param thread` in the javadoc.
```
@return UUID uncaught exception's UUID.
```

File: sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java [hunk 6]
```diff
@@ -1097,14 +1097,16 @@ void setLogSerializer(LogSerializer logSerializer) {
...
      * @param throwable uncaught exception or error.
      */
-    void saveUncaughtException(Thread thread, Throwable throwable) {
+    public UUID saveUncaughtException(Thread thread, Throwable throwable) {
```
Comment (guperrot) [inline sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java:1100]: We do not have public methods at instance level. All public APIs are meant to be accessed from static and be documented.

Comment (ShreyaA) [reply sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java:1100]: I see. Then perhaps I could introduce a public static method, (let's call it saveExceptionReturnReportUUID) that will call UUID saveUncaughtException(Thread thread, Throwable throwable) internally and return the UUID. Then I'll call saveExceptionReturnReportUUID in my custom exception handler.

Will that be fine?

Comment (MatkovIvan) [reply sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java:1100]: After a discussion with the team, we decided not to introduce a new static API for this. However, we agreed to merge this PR with the explicit note that this method can be changed in future versions without warning.

File: sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java [hunk 7]
```diff
@@ -1097,14 +1097,16 @@ void setLogSerializer(LogSerializer logSerializer) {
      * @param thread    thread where crash occurred.
      * @param throwable uncaught exception or error.
```
Comment (MatkovIvan) [inline sdk/appcenter-crashes/src/main/java/com/microsoft/appcenter/crashes/Crashes.java:1098]:
```suggestion
     * @param throwable uncaught exception or error.
     * @return UUID uncaught exception's UUID.
```

**Review by MatkovIvan — APPROVED** [review]

**Pull request merged**